ambry-router/src/main/java/com.github.ambry.router/GetOperation.java

```diff
@@ -0,0 +1,215 @@
+   * Return the {@link FutureResult} associated with this operation.
+   * @return the {@link FutureResult} associated with this operation.
+   */
+  FutureResult<T> getFuture() {
+    return operationFuture;
+  }
+
+  /**
+   * Return the {@link Callback} associated with this operation.
+   * @return the {@link Callback} associated with this operation.
+   */
+  Callback<T> getCallback() {
+    return operationCallback;
+  }
+
+  /**
+   * The exception associated with this operation if it failed; null otherwise.
+   * @return exception associated with this operation if it failed; null otherwise.
+   */
+  Exception getOperationException() {
+    return operationException;
+  }
+
+  /**
+   * Return the result of the operation.
+   * @return the operation result.
+   */
+  T getOperationResult() {
+    return operationResult;
+  }
+
+  /**
+   * Return the blob id string
+   * @return return the blob id string
+   */
+  String getBlobIdStr() {
+    return blobId.getID();
+  }
+
+  /**
+   * returns whether the operation has completed.
+   * @return whether the operation has completed.
+   */
+  boolean isOperationComplete() {
+    return operationCompleted;
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   * @return the {@link MessageFormatFlags} to associate with the requests for this operation.
+   */
+  abstract MessageFormatFlags getOperationFlag();
+
+  /**
+   * For this operation, create and populate get requests to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  abstract void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback);
+
+  /**
+   * Handle the given {@link ResponseInfo} received for a request that was sent out.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  abstract void handleResponse(ResponseInfo responseInfo);
+
+  /**
+   * Process the given {@link ServerErrorCode} and set operation status accordingly.
+   * @param errorCode the {@link ServerErrorCode} to process.
+   */
+  void processServerError(ServerErrorCode errorCode) {
+    switch (errorCode) {
+      case Blob_Deleted:
+        logger.trace("Requested blob was deleted");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDeleted));
+        break;
+      case Blob_Expired:
+        logger.trace("Requested blob is expired");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobExpired));
+        break;
+      case Blob_Not_Found:
+        logger.trace("Requested blob was not found on this server");
+        setOperationException(new RouterException("Server returned: " + errorCode, RouterErrorCode.BlobDoesNotExist));
+        break;
+      default:
+        setOperationException(
```

Comment:
If that happens, it wouldn't immediately return though. It would continue to look for the blob in other replicas, and maybe it will find it somewhere or come to know that it is deleted or expired. The situation we are talking about only arises if some replicas returned `Blob_Not_Found` and all the others return `Unexpected_Error` or whatever. When this happens, isn't it clealry better to return `Blob_Not_Found`? There is no real way of knowing that had there been no IO issue, the blob would be successfully obtained. There is a good chance that the blob is not there at all since some of the other replicas said that it is not found.

In your scheme, even if one replica out of *N* is down, and the other *N-1* got back with `Blob_Not_Found`, you are suggesting we should return an `UnexpectedInternalError` - which is clearly not ideal.